kernel/src/transaction.rs

```diff
@@ -0,0 +1,543 @@
+use std::iter;
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use crate::actions::get_log_commit_info_schema;
+use crate::actions::COMMIT_INFO_NAME;
+use crate::error::Error;
+use crate::expressions::{column_expr, column_name, Scalar, StructData};
+use crate::path::ParsedLogPath;
+use crate::schema::{MapType, StructField, StructType};
+use crate::snapshot::Snapshot;
+use crate::{DataType, DeltaResult, Engine, EngineData, Expression, Version};
+
+const KERNEL_VERSION: &str = env!("CARGO_PKG_VERSION");
+const UNKNOWN_OPERATION: &str = "UNKNOWN";
+
+/// A transaction represents an in-progress write to a table. After creating a transaction, changes
+/// to the table may be staged via the transaction methods before calling `commit` to commit the
+/// changes to the table.
+///
+/// # Examples
+///
+/// ```rust,ignore
+/// // create a transaction
+/// let mut txn = table.new_transaction(&engine)?;
+/// // stage table changes (right now only commit info)
+/// txn.commit_info(Box::new(ArrowEngineData::new(engine_commit_info)));
+/// // commit! (consume the transaction)
+/// txn.commit(&engine)?;
+/// ```
+pub struct Transaction {
+    read_snapshot: Arc<Snapshot>,
+    operation: Option<String>,
+    commit_info: Option<Arc<dyn EngineData>>,
+}
+
+impl std::fmt::Debug for Transaction {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&format!(
+            "Transaction {{ read_snapshot version: {}, commit_info: {} }}",
+            self.read_snapshot.version(),
+            self.commit_info.is_some()
+        ))
+    }
+}
+
+impl Transaction {
+    /// Create a new transaction from a snapshot. The snapshot will be used to read the current
+    /// state of the table (e.g. to read the current version).
+    ///
+    /// Instead of using this API, the more typical (user-facing) API is
+    /// [Table::new_transaction](crate::table::Table::new_transaction) to create a transaction from
+    /// a table automatically backed by the latest snapshot.
+    pub(crate) fn new(snapshot: impl Into<Arc<Snapshot>>) -> Self {
+        Transaction {
+            read_snapshot: snapshot.into(),
+            operation: None,
+            commit_info: None,
+        }
+    }
+
+    /// Consume the transaction and commit it to the table. The result is a [CommitResult] which
+    /// will include the failed transaction in case of a conflict so the user can retry.
+    pub fn commit(self, engine: &dyn Engine) -> DeltaResult<CommitResult> {
+        // step one: construct the iterator of actions we want to commit
+        // note: only support commit_info right now (and it's required)
+        let engine_commit_info = self
+            .commit_info
+            .clone()
+            .ok_or_else(|| Error::MissingCommitInfo)?;
```

Comment:
```suggestion
        let engine_commit_info = self
            .commit_info
            .as_ref()
            .ok_or_else(|| Error::MissingCommitInfo)?;
        let engine_commit_info = engine_commit_info.as_ref();
```

Then below your `generate_commit_info` becomes:
```rust
fn generate_commit_info(
    engine: &dyn Engine,
    operation: Option<&str>,
    engine_commit_info: &dyn EngineData,
) -> DeltaResult<Box<dyn EngineData>> {
```

And finally you don't need `as_ref` in the evaluate call in that function:
```rust
    commit_info_evaluator.evaluate(engine_commit_info)
```

Reply:
nice - got it all fixed up thanks!!